cordapp/src/main/kotlin/net/corda/plugins/CordappPlugin.kt

```diff
@@ -76,15 +79,53 @@ class CordappPlugin : Plugin<Project> {
         }
         jarTask.dependsOn(task)
 
-        val signTask = project.task("signCordappJar")
+        val optionalSignTask = createSigningTask(project, "signCordapp", false)
+        jarTask.finalizedBy(optionalSignTask)
+
+        val signTask = project.tasks.findByName("sourceJar") //task in corda-publish-utils plugin
+        if (signTask != null ) {
+            val signTaskForPublishingTask = createSigningTask(project, "signCordappForPublishing", true)
+            val jarPublishingTask = signTask as Jar
+            jarPublishingTask.dependsOn(signTaskForPublishingTask)
+        }
+    }
+
+    private fun createSigningTask(project: Project, taskName: String, warnOnDevSetup: Boolean) : Task {
+        val signTask = project.task(taskName)
```

Comment:
If there's any sensible lines down which this function can be decomposed please do.